src/librustc_lint/builtin.rs

```diff
@@ -2153,123 +2153,208 @@ impl ClashingExternDeclarations {
+                SeenSetResult::Computed(result) => result,
+                // We are already in the process of computing structural sameness for this type,
+                // meaning we've found a cycle. The types are structurally same, then.
+                SeenSetResult::Computing => true,
+                // We haven't seen this combination of types at all -- compute their sameness.
+                SeenSetResult::Unseen => {
+                    seen_types.mark_computing(a, b);
+                    let tcx = cx.tcx;
+                    let result = if a == b || rustc_middle::ty::TyS::same_type(a, b) {
+                        // All nominally-same types are structurally same, too.
+                        true
+                    } else {
+                        // Do a full, depth-first comparison between the two.
+                        use rustc_middle::ty::TyKind::*;
+                        let a_kind = &a.kind;
+                        let b_kind = &b.kind;
+
+                        let compare_layouts = |a, b| -> bool {
+                            let a_layout = &cx.layout_of(a).unwrap().layout.abi;
+                            let b_layout = &cx.layout_of(b).unwrap().layout.abi;
+                            debug!("{:?} == {:?} = {}", a_layout, b_layout, a_layout == b_layout);
+                            a_layout == b_layout
+                        };
+
+                        #[allow(rustc::usage_of_ty_tykind)]
+                        let is_primitive_or_pointer = |kind: &ty::TyKind<'_>| {
+                            kind.is_primitive() || matches!(kind, RawPtr(..))
+                        };
 
-            #[allow(rustc::usage_of_ty_tykind)]
-            let is_primitive_or_pointer =
-                |kind: &ty::TyKind<'_>| kind.is_primitive() || matches!(kind, RawPtr(..));
-
-            match (a_kind, b_kind) {
-                (Adt(_, a_substs), Adt(_, b_substs)) => {
-                    let a = a.subst(cx.tcx, a_substs);
-                    let b = b.subst(cx.tcx, b_substs);
-                    debug!("Comparing {:?} and {:?}", a, b);
-
-                    if let (Adt(a_def, ..), Adt(b_def, ..)) = (&a.kind, &b.kind) {
-                        // Grab a flattened representation of all fields.
-                        let a_fields = a_def.variants.iter().flat_map(|v| v.fields.iter());
-                        let b_fields = b_def.variants.iter().flat_map(|v| v.fields.iter());
-                        compare_layouts(a, b)
+                        match (a_kind, b_kind) {
+                            (Adt(_, a_substs), Adt(_, b_substs)) => {
+                                let a = a.subst(cx.tcx, a_substs);
+                                let b = b.subst(cx.tcx, b_substs);
+                                debug!("Comparing {:?} and {:?}", a, b);
+
+                                if let (Adt(a_def, ..), Adt(b_def, ..)) = (&a.kind, &b.kind) {
```

Comment:
why aren't you directly taking the def from the match statement here? `subst` doesn't change `DefId`s, does it?